rosplan_planning_system/src/PlannerInterface/LPGPlannerInterface.cpp

```diff
@@ -0,0 +1,141 @@
+#include "rosplan_planning_system/PlannerInterface/LPGPlannerInterface.h"
+
+namespace KCL_rosplan {
+
+	/*-------------*/
+	/* constructor */
+	/*-------------*/
+
+	LPGPlannerInterface::LPGPlannerInterface(ros::NodeHandle& nh)
+	{
+		node_handle = &nh;
+
+		plan_server = new actionlib::SimpleActionServer<rosplan_dispatch_msgs::PlanAction>((*node_handle), "start_planning", boost::bind(&PlannerInterface::runPlanningServerAction, this, _1), false);
+
+		// publishing raw planner output
+		std::string plannerTopic = "planner_output";
+		node_handle->getParam("planner_topic", plannerTopic);
+		plan_publisher = node_handle->advertise<std_msgs::String>(plannerTopic, 1, true);
+
+		// start planning action server
+		plan_server->start();
+	}
+
+	LPGPlannerInterface::~LPGPlannerInterface()
+	{
+		delete plan_server;
+	}
+
+	/**
+	 * Runs external commands
+	 */
+	std::string LPGPlannerInterface::runCommand(std::string cmd) {
+		std::string data;
+		FILE *stream;
+		char buffer[1000];
+		stream = popen(cmd.c_str(), "r");
+		while ( fgets(buffer, 1000, stream) != NULL )
+			data.append(buffer);
+		pclose(stream);
+		return data;
+	}
+
+	/*------------------*/
+	/* Plan and process */
+	/*------------------*/
+
+	/**
+	 * passes the problem to the Planner; the plan to post-processing.
+	 */
+	bool LPGPlannerInterface::runPlanner() {
+
+		// save problem to file for LPG
+		if(use_problem_topic && problem_instance_recieved) {
+			ROS_INFO("KCL: (%s) (%s) Writing problem to file.", ros::this_node::getName().c_str(), problem_name.c_str());
+			std::ofstream dest;
+			dest.open((problem_path).c_str());
+			dest << problem_instance;
+			dest.close();
+		}
+
+		// prepare the planner command line
+		std::string str = planner_command;
+		std::size_t dit = str.find("DOMAIN");
+		if(dit!=std::string::npos) str.replace(dit,6,domain_path);
+		std::size_t pit = str.find("PROBLEM");
+		if(pit!=std::string::npos) str.replace(pit,7,problem_path);
+        std::string updatePlan = "cp "+data_path+"../../rosplan_planning_system/common/bin/LPG-td-1.0/lpgplan.SOL"+" "+data_path+"plan.pddl";
```

Comment:
This works.
I wonder what will happen when someone uses the planner in a different directory (which they can do using the command line parameter.) The solution needs to be in the same directory as the planner, which could be difficult to extract from the command line if e.g., there are multiple paths in the command line, and the version of LPG is not 1.0.

I've made a note to revisit this at some later date.

Reply:
hmm, actually LPG is not dependend on the folder (only TFD is), will fix. However, the output of the plan in the command line must be at a specific location (will use the current plan path  $(find rosplan_planning_system)/common/ as the plan output).

Ex command:
       <arg name="planner_command" value="(cd $(find rosplan_planning_system)/common/bin/; timeout 10 ./lpg-td-1.0 -o DOMAIN -f PROBLEM -speed -out $(find rosplan_planning_system)/common/lpgplan)" />

Reply:
It's a tricky one, remembering that the data_path is also a parameter. Documentation can advise the user on where to put things.